unbricking.md

```diff
@@ -0,0 +1,193 @@
+# G-010S-A Unbricking
+
+If anything goes wrong during an update the SFP module can be unbricked through
+a serial console.
+
+## Preparation / Unlock Bootloader
+<a name="preparation"></a>
+
+By default the bootloader is locked and will not allow interrupting the boot.
+Unlock the bootloader while the SFP module is still working using:
+```
+fw_setenv asc0 0            # Fully activate console
+fw_setenv bootdelay 5       # Prompt for access to u-boot console
+fw_setenv preboot           # Avoid console hanging while initializing network
+```
+
+## Serial Console Cable
+
+The serial console can be accessed with a UART to SFP cable. One tested UART
+board is Diligent PmodUSBUART. Ideally the UART board provides 3.3V for the SFP
+module. The SFP connector (e.g., Molex 74441-0010) is connected to the UART on
+4 pins:
+| UART Pin     | SFP Pin  |
+| ------------ | ---------|
+| TX           | 3        |
+| RX           | 6        |
+| GND          | 10       |
+| 3.3V         | 15,16    |
+
+One option is to solder it to a universal printed circuit board or similar.
+Another option that worked well is to solder flat band cables straight to the
+connector and then glue the connector to a surface board with epoxy glue.
+
+The 3.3V connection from the UART will likely provide enough power to get to
+u-boot and through the early kernel boot. It will likely not be enough to fully
+boot up and activate the GPON and ethernet interface. Reaching the interface
+initialization will then cause a restart of the module.
+
+## Connecting to the Console
+
+Connect to the UART with 115200 baud, e.g.:
+```
+screen /dev/tty.usb-device 115200
+```
+
+Then connect the SFP module to the Molex connector.  If the bootloader is
+unlocked the boot prompt looks like this:
+```
+ROM: V1.1.4
+ROM: CFG 0x00000006
+ROM: SFLASH-4
+Chip:  FALCON-S (A22)
+Bootmode: 0x06
+Reset cause: Power-On Reset
+CPU Clock: 400 MHz
+Net:   SGMII, SERDES [PRIME]
+Press SPACE to delay and Ctrl-C to abort autoboot in 5 seconds
+FALCON =>
+```
+Press Ctrl-C to abort the boot to get u-boot access.
+
+If the bootloader is still locked there is no option to interrupt the boot (or even this level of boot message). In that case see [Unbrick
+Unprepared SFP Modules](#unlock_bricked).
+
+## Fixing Environment Issues
+
+If the bricking was caused by a bad u-boot environment it can be fixed using
+``setenv``. After changing the environment use ``saveenv`` to make the changes
+permanent.
+
+For example, if the active image 1 is not bootable, but image 0 is still
+bootable then use:
+```
+setenv commit 0
+saveenv
+```
+
+## Replacing Firmware
+
+Credit to anon23891239 describing it [here](https://forum.openwrt.org/t/support-ma5671a-sfp-gpon/48042/24).
+
+If no bootable firmware image is left the images can be replaced via u-boot.
+The two images are stored at 0xC0000(length 0x740000) and 0x800000(length
```

Comment:
These are values from other Lantiq SFPs.

`mtdparts=sflash:256k(uboot)ro,512k(uboot_env),6144k(image0),6144k(linux),1024k(configfs),2112k(logfs),64k(ri),64k(sfp),64k(ribackup)`
So :
 - image0 is at 0xC0000, length 0x600000
 - image1 is at 0x6C0000, length 0x600000

Reply:
done